cdap-app-templates/cdap-etl/cdap-etl-core/src/test/java/io/cdap/cdap/etl/common/OAuthMacroEvaluatorTest.java

```diff
@@ -0,0 +1,139 @@
+
+  private static NettyHttpService httpService;
+  private static ServiceDiscoverer serviceDiscoverer;
+
+  @BeforeClass
+  public static void init() throws Exception {
+    httpService = NettyHttpService.builder("OAuthTest")
+      .setHttpHandlers(new OAuthHandler(
+        ImmutableMap.of(
+          PROVIDER, ImmutableMap.of(CREDENTIAL_ID, new OAuthInfo("accessToken", "bearer"))
+        )))
+      .build();
+
+    httpService.start();
+
+    InMemoryDiscoveryService discoveryService = new InMemoryDiscoveryService();
+    String discoveryName = ServiceDiscoverable.getName(NamespaceId.SYSTEM.getNamespace(),
+                                                       Constants.PIPELINEID, ProgramType.SERVICE,
+                                                       Constants.STUDIO_SERVICE_NAME);
+    discoveryService.register(new Discoverable(discoveryName, httpService.getBindAddress()));
+    serviceDiscoverer = new AbstractServiceDiscoverer(NamespaceId.DEFAULT.app("testapp").spark("testspark")) {
+
+      @Override
+      protected DiscoveryServiceClient getDiscoveryServiceClient() {
+        return discoveryService;
+      }
+    };
+  }
+
+  @AfterClass
+  public static void finish() throws Exception {
+    httpService.stop();
+  }
+
+  @Test
+  public void testOAuthMacro() {
+    MacroEvaluator macroEvaluator = new OAuthMacroEvaluator(serviceDiscoverer);
+    String value = macroEvaluator.evaluate(OAuthMacroEvaluator.FUNCTION_NAME, PROVIDER, CREDENTIAL_ID);
+    OAuthInfo oAuthInfo = GSON.fromJson(value, OAuthInfo.class);
+
+    Assert.assertEquals("accessToken", oAuthInfo.accessToken);
+    Assert.assertEquals("bearer", oAuthInfo.tokenType);
+  }
+
+  /**
+   * A http handler to provide the OAuth endpoint.
+   */
+  public static final class OAuthHandler extends AbstractHttpHandler {
+
+    private final Map<String, Map<String, OAuthInfo>> credentials;
+
+    public OAuthHandler(Map<String, Map<String, OAuthInfo>> credentials) {
+      this.credentials = credentials;
+    }
+
+    @Path("/v3/namespaces/system/apps/" + Constants.PIPELINEID + "/services/" +
+      Constants.STUDIO_SERVICE_NAME + "/methods/v1/oauth/provider/{provider}/credential/{credentialId}")
+    @GET
+    public void getOAuth(HttpRequest request, HttpResponder responder,
+                         @PathParam("provider") String provider,
+                         @PathParam("credentialId") String credentialId) {
+      Map<String, OAuthInfo> providerCredentials = credentials.get(provider);
+      if (providerCredentials == null) {
+        responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+        return;
+      }
+
+      OAuthInfo oAuthInfo = providerCredentials.get(credentialId);
+      if (oAuthInfo == null) {
+        responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+        return;
+      }
+
+      responder.sendString(HttpResponseStatus.OK, GSON.toJson(oAuthInfo));
+    }
+  }
+
+  private static final class OAuthInfo {
```

Comment:
The actual handler returns an json string. From the OAuthMacroEvaluator point of view, it always evaluate value as a string. It is the PluginInstantiator that will turn the string into object based on the plugin config field type.

Reply:
Ah I see, I didn't put those two together. So the intended usage of this feature is to have a field in the plugin config of type `GetAccessTokenResponse` and then put the macro in there. What is the use case for making it a config field + macro instead of just an extension to the plugin config? Is it to allow the user to manually input OAuth access tokens (e.g. for testing)?

Reply:
It is intended to be used as a plugin config field. You can refer to https://github.com/data-integrations/salesforce/pull/61/files#diff-eaa9626ca9c399947096e3137f754895fb2e88ce2f2efe1cf53e231dc7d58db2R42 for the actual usage.

Reply:
Understood, thanks!